compat/compat_mman.c

```diff
@@ -0,0 +1,279 @@
+/*
+ * compat/compat_mman.h
+ *
+ * Copyright (C) 2013   JP Ikaheimonen <jp_ikaheimonen@mentor.com>
+ *               2016   Michael Jeanson <mjeanson@efficios.com>
+ *
+ * These sources are based on ftp://g.oswego.edu/pub/misc/malloc.c
+ * file by Doug Lea, released to the public domain.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#ifdef __MINGW32__
+
+#include <assert.h>
+#include <errno.h>
+#include <io.h>
+#include <pthread.h>
+#include <stdlib.h>
+#include <windows.h>
+#include <babeltrace/compat/mman-internal.h>
+
+struct mmap_info {
+	HANDLE file_handle; /* the duplicated handle */
+	HANDLE map_handle;  /* handle returned by CreateFileMapping */
+	void *start;        /* ptr returned by MapViewOfFile */
+};
+
+/*
+ * This mutex protects the array of memory mappings and its associated
+ * counters. (mmap_infos, mmap_infos_cur mmap_infos_max)
+ */
+static pthread_mutex_t mmap_mutex = PTHREAD_MUTEX_INITIALIZER;
+
+static int mmap_infos_cur = 0;
+static int mmap_infos_max = 0;
+static struct mmap_info *mmap_infos = NULL;
```

Comment:
We have GLib. Why not using a protected global GHashTable instead of this manually allocated and reallocated array? This would feel more Babeltracy since it's an approach we use at several places in the library and plugins.